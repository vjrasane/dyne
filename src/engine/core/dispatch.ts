import { exists } from "../utils";
import Effect from "../effects/effect";

export type Dispatch = (msg: object) => void;

type Queues = {
  msg: object[];
  effect: Effect<any>[];
};

export const queues: Queues = {
  msg: [],
  effect: []
};

const dispatcher = <A>(queue: A[], queueable: A): void => {
  exists(queueable) && queue.push(queueable);
};

export const dispatch = (queueable: Effect<any> | object): void => {
  queueable instanceof Effect
    ? dispatcher(queues.effect, queueable)
    : dispatcher(queues.msg, queueable);
};
