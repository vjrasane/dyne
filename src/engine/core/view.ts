import freeze from "deep-freeze";
import { isFunction, always } from "../../utils";
import { VirtualDom } from "../render";
import { exists } from "../../utils";

export type View<M> = (model: M) => VirtualDom | VirtualDom;

export type Viewer<M> = View<M>;

export const getViewer = <M>(view: View<M>): Viewer<M> => {
  if (!exists(view)) {
    throw new Error(
      `Expected 'view' to be a function or element, was: '${view}'`
    );
  } else if (isFunction(view)) {
    return model => freeze(view(model));
  } else {
    return always(freeze(view));
  }
};
