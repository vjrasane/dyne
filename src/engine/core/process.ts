import freeze from "deep-freeze";
import { exists } from "../utils";
import Return from "../effects/return";
import { Updated, Update } from "./update";
import { dispatch } from "../core/dispatch";

export default <A>(updated: Updated<A>, previous?: A): A => {
  const processed: Return<A> = Update.wrap(updated);
  processed.effects.forEach(dispatch);
  return exists(processed.model) ? freeze(processed.model) : previous;
};
