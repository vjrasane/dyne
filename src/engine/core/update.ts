import { utils as U } from "dyne-commons";
import Return, { Effectable } from "../effects/return";
import Effect from "../effects/effect";
import { CmdFx } from "../effects/types";
import process from "./process";
import { Dispatch } from "./dispatch";

const throwUpdateError = (update: any): (() => never) =>
  U.throwError(
    new Error(
      `Attempted to pass message to 'update', but it was not a function, was: ${typeof update}`
    )
  );

export type Updated<A> = Return<A> | Effect<any> | CmdFx<A> | A;

export type Updater<A> = (msg: object, model: A) => A;

export type Update<A> = (msg: object, model: A) => Updated<A>;

export const Update = <A>(model: A, ...effects: Effectable<any>[]): Return<A> =>
  new Return(model, ...effects);

Update.wrap = <A>(wrappable: Updated<A>): Return<A> => {
  if (wrappable instanceof Return) {
    return wrappable;
  } else if (
    U.isFunction(wrappable) ||
    wrappable instanceof Promise ||
    wrappable instanceof Effect
  ) {
    return Update(null, wrappable);
  } else {
    return Update(<A>wrappable);
  }
};

export const getUpdater = <M>(
  dispatch: Dispatch,
  update: Update<M>
): Updater<M> => {
  if (U.isFunction(update)) {
    return (msg: object, model: M) =>
      process(dispatch, update(msg, model), model);
  }
  return throwUpdateError(update);
};
