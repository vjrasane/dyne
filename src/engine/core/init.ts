import process from "../core/process";
import { Updated } from "./update";
import { isFunction, exists, always } from "../../utils";

export type Init<F, M> = (flags: F) => Updated<M>;

export type Initializer<F, M> = (flags: F) => M;

export const getInitializer = <F, M>(init?: Init<F, M>): Initializer<F, M> => {
  if (!exists(init)) return always();
  if (!isFunction(init)) {
    throw new Error(`Given 'init' was not a function, was: ${typeof init}`);
  }
  return (flags: F): M => process(init(flags));
};
