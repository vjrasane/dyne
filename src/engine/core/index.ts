import { renderer, VirtualDom } from "../render";
import { getInitializer, Initializer, Init } from "../core/init";
import { getUpdater, Updater, Update } from "../core/update";
import { getViewer, View, Viewer } from "../core/view";
import { dispatch, queues } from "../core/dispatch";
import { Sink } from "../../utils";

const UPDATE_INTERVAL = 1; /* milliseconds */
const VIEW_INTERVAL = 10; /* milliseconds */
const RENDER_INTERVAL = 10; /* milliseconds */
const EFFECT_INTERVAL = 1; /* milliseconds */

const loop = (procedure: Sink, interval: number): void => {
  let lock = false;
  setInterval((): void => {
    if (!lock) {
      lock = true;
      procedure();
      lock = false;
    }
  }, interval);
};

type Setup<M> = {
  init: Init<M>;
  update: Update<M>;
  view: View<M>;
};

const engine = <M>({ init, update, view }: Setup<M>) => {
  const updater: Updater<M> = getUpdater(update);
  const viewer: Viewer<M> = getViewer(view);
  const initializer: Initializer<M> = getInitializer(init);

  return ({ node, flags }) => {
    let model: M = initializer(flags);

    // update loop
    loop(() => {
      if (queues.msg.length > 0) {
        model = updater(queues.msg.shift(), model);
      }
    }, UPDATE_INTERVAL);

    // view loop
    let initialRender: Boolean = false;
    let viewedModel: M;
    let virtualDom: VirtualDom;
    const render = () => {
      virtualDom = viewer(model);
      viewedModel = model;
    };

    loop(() => {
      if (!initialRender) {
        render();
        initialRender = true;
      } else if (model && model !== viewedModel) {
        render();
      }
    }, VIEW_INTERVAL);

    // render loop
    let renderedVirtualDom;
    loop(() => {
      if (virtualDom && virtualDom !== renderedVirtualDom) {
        renderer(node, virtualDom, renderedVirtualDom);
        renderedVirtualDom = virtualDom;
      }
    }, RENDER_INTERVAL);

    // effect loop
    loop(() => {
      if (queues.effect.length > 0) {
        const effect = queues.effect.shift();
        // execute effect with the message dispatcher
        effect.execute(dispatch);
      }
    }, EFFECT_INTERVAL);
  };
};

export default engine;
