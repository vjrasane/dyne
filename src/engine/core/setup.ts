import { renderer, VirtualDom } from "../render";
import { getInitializer, Initializer, Init } from "./init";
import { getUpdater, Updater, Update } from "./update";
import { getViewer, View, Viewer } from "./view";
import { dispatch, queues } from "./dispatch";

export type Setup<M> = {
  init: Init<M>;
  update: Update<M>;
  view: View<M>;
};

export type Opts = {
  node: HTMLElement;
  flags: object;
};

export const setup = <M>({ init, update, view }: Setup<M>) => {
  const updater: Updater<M> = getUpdater(update);
  const viewer: Viewer<M> = getViewer(view);
  const initializer: Initializer<M> = getInitializer(init);

  return ({ node, flags }: Opts) => {
    let model: M = initializer(flags);

    const updateProcedure = () => {
      if (queues.msg.length > 0) {
        model = updater(queues.msg.shift(), model);
      }
    };

    let viewedModel: M;
    let virtualDom: VirtualDom;
    const viewVirtualDom = () => {
      virtualDom = viewer(model);
      viewedModel = model;
    };

    // initial dom
    viewVirtualDom();
    const viewProcedure = () => {
      if (model && model !== viewedModel) {
        viewVirtualDom();
      }
    };

    let renderedVirtualDom;
    const renderProcedure = () => {
      if (virtualDom && virtualDom !== renderedVirtualDom) {
        renderer(node, virtualDom, renderedVirtualDom);
        renderedVirtualDom = virtualDom;
      }
    };

    const effectProcedure = () => {
      if (queues.effect.length > 0) {
        const effect = queues.effect.shift();
        // execute effect with the message dispatcher
        effect.execute(dispatch);
      }
    };

    return {
      update: updateProcedure,
      view: viewProcedure,
      render: renderProcedure,
      effect: effectProcedure
    };
  };
};
