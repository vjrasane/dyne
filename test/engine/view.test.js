import Dyne from "~";
import { getViewer } from "~/engine/core/view";

describe("view", () => {
  it("returns basic virtual dom", () => {
    const viewer = getViewer(() => <div>HELLO WORLD!</div>);
    const virtualDom = viewer();
    expect(virtualDom).toMatchSnapshot();
  });

  it("throws error with undefined view", () => {
    expect(getViewer).toThrowError();
  });

  it("throws error with object view", () => {
    expect(() => getViewer({})).toThrowError();
  });
});
