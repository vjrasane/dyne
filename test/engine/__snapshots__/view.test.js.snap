// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view returns basic virtual dom 1`] = `
DyneElement {
  "children": Array [
    "HELLO WORLD!",
  ],
  "props": Array [],
  "type": "div",
}
`;
